### PR TITLE
feat[file]: add s3 functionality

### PR DESCRIPTION
**Pull request by m-mattia-m**
- Configure S3 endpoint, bucket, base path, credentials
- Upload file to S3
- Get file from S3
- delete file from S3

PS: Please bear in mind that I've never used NestJS before, so I'm not sure if everything is best practice, but it works. Don't hesitate to tell me what I should change or update.

### CODE REVIEW

**Comment by stonith404** [comment]
Thank you @m-mattia-m :) I’ll try to review the PR as soon as I can. Since I’ll be on vacation next week and have a lot going on, it may take me some time to review and test it fully.

**Review by stonith404 — CHANGES_REQUESTED** [review]
Sorry for my late response, I have much going on.

I noticed an issue with how chunked file uploads are being handled. Currently, only the first chunk of a file is successfully uploaded to S3. Pingvin Share relies on chunked uploads, where the frontend splits a file into multiple parts and uploads each part sequentially. 

With your current implementation, only the initial chunk is uploaded. You can verify this by trying to upload a file larger than the chunk size (default is 10 MB).

**Comment by stonith404** [comment]
Additionally, I would suggest to create a separate S3 file service as the methods in the file service get a bit too complicated when both S3 and local file storage are handled.

**Comment by m-mattia-m** [comment]
No worries :)

I can see what you mean by "a bit too complicated" :D. I will create a separate service for the S3 functionality. Yes you are right, large files are failing during the upload. I've never tested it with large files which I should have done. I will implement this feature in a few weeks but I have also much going on :).

**Comment by stonith404** [comment]
@m-mattia-m Sure, take your time :)

**Comment by m-mattia-m** [comment]
FYI: I fixed the chunking problem and split the file service into a local and a S3 service. Consider that I need to refactor these changes, but I will do that in a few days. If you have time, I would appreciate it when you could test the changes if you can find any issues.

**Comment by m-mattia-m** [comment]
@stonith404 I updated the code a bit, and I hope it's fine. Don't hesitate to tell me what I should change or update. :)

**Review by stonith404 — CHANGES_REQUESTED** [review]
Thanks for taking the time to implement my feedback :)

Most things are now working as expected, but I’ve noticed a few remaining issues. Please take a look at my additional comments.

File: backend/src/file/file.service.ts [hunk 1]
```diff
@@ -1,162 +1,66 @@
...
+  metaData: {
+    id: string,
+    size: string,
+    createdAt: Date
+    mimeType: string | false,
+    name: string,
+    shareId: string
+  }
+  file: Readable
+}
```
[the remark this answers is not in the document]
Comment (m-mattia-m) [reply backend/src/file/file.service.ts:66]: When I remove this type, I get errors like `Property metaData does not exist on type File'.

File: backend/src/file/s3.service.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+@Injectable()
+export class S3FileService {
+  private s3: S3Client;
+  private multipartUploads: Record<string, {
+    uploadId: string;
+    parts: Array<{ ETag: string | undefined; PartNumber: number }>
+  }> = {};
+
+  constructor(private prisma: PrismaService, private config: ConfigService) {
+    this.s3 = new S3Client({
```
[the remark this answers is not in the document]
Comment (m-mattia-m) [reply backend/src/file/s3.service.ts]: Yes, of course. I had this, but I probably lost it when merging the changes from the main branch. 😅 However, I reimplemented it.

**Comment by m-mattia-m** [comment]
I noticed two things that I will change/implement in a few weeks when I have more time. 
- When you try to open the preview for a PDF that is larger than 10Mb (chunk size), an error occurs (it downloads the file (without the correct name and type extension) instead of displaying the content in the PDF viewer).
- If you change the storage provider during runtime, the backend searches in the new location where it can't find these files that were uploaded before the storage change, so I will add a location: `s3`, `local` to each file.

File: backend/src/file/file.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (stonith404) [reply backend/src/file/file.service.ts:66]: Oh I see, makes sense.

**Comment by stonith404** [comment]
Thanks for the changes!

Sure, take your time. I think it would be sufficient to add the provider information to the share, instead of adding it to every file. 

When these two issues are fixed it should finally be ready to merge :)

**Comment by m-mattia-m** [comment]
I fixed the two issues I mentioned before. Now everything should work :).

**Review by stonith404 — APPROVED** [review]
I made some minor changes like formatting the code, adding the Prisma migration and removing `storageProvider` from the dto as the client doesn't need this property.

For me it looks good now. Should I merge it?

**Comment by m-mattia-m** [comment]
That would be great. :)

**Comment by stonith404** [comment]
Great, thanks again for your contribution!

**Comment by GitNees** [comment]
hi @m-mattia-m I just filed a bug report as I can use S3 buckets and upload, but impossible to delete the files from the shares.

**Comment by m-mattia-m** [comment]
#705

**Pull request merged**